### PR TITLE
Fix: Object variable names can be used as formal arguments of methods other than constructors

### LINKED ISSUE #13152: Initializing formal parameters are available to all methods
### Steps to reproduce

Is it intended that this works?

```vim
vim9script

class Foo
  this.foo: number
  def SetFoo(this.foo)
  enddef
  def GetFoo(): number
    return this.foo
  enddef
endclass

var f = Foo.new()
f.SetFoo(42)
echomsg f.GetFoo()
```

### Expected behaviour

I don't feel strongly about it one way or the other but allowing this feature is normally (from a brief survey of other languages) only available in constructors.

### Version of Vim

9.0.1927

### Environment

Ubuntu 20.04

### Logs and stack traces

_No response_

### PR DESCRIPTION
**Pull request by h-east**
Close: #13152

### CODE REVIEW

File: src/userfunc.c [hunk 1]
```diff
@@ -320,6 +320,16 @@ get_function_args(
 		++p;
 	    char_u *argend = p;
 
+	    // object variable can use constructor
```
Comment (yegappan) [inline src/userfunc.c]: Can you update this comment to "object variable this.<name> can be used only in a constructor"?

File: src/testdir/test_vim9_class.vim [hunk 2]
```diff
@@ -955,6 +955,28 @@ def Test_class_new_with_object_member()
   END
   v9.CheckSourceFailure(lines, 'E1013:')
 
+  lines =<< trim END
+      vim9script
```
Comment (yegappan) [inline src/testdir/test_vim9_class.vim]: The PR https://github.com/vim/vim/pull/13206 will adjust the indentation to use two spaces in a heredoc.  Can you use two spaces for indentation in the heredoc?

File: src/testdir/test_vim9_class.vim [hunk 3]
```diff
@@ -955,6 +955,28 @@ def Test_class_new_with_object_member()
...
+      def Check()
+        try
+          var c = C.new()
+          c.MethodA('abc')
+        catch
+          assert_report($'Unexpected exception was caught: {v:exception}')
+        endtry
+      enddef
+
+      Check()
```
Comment (yegappan) [inline src/testdir/test_vim9_class.vim]: As the compilation of the class will fail, will it compile the Check() function and invoke it?  Is this function needed in this test?

File: src/errors.h [hunk 4]
```diff
@@ -3529,6 +3529,8 @@ EXTERN char e_public_keyword_not_supported_for_method[]
 	INIT(= N_("E1388: Public keyword not supported for a method"));
 EXTERN char e_missing_name_after_implements[]
 	INIT(= N_("E1389: Missing name after implements"));
+EXTERN char e_cannot_use_a_object_variable_except_with_the_new_method_str[]
+	INIT(= N_("E1390: Cannot use a object variable \"this.%s\" except with the \"new\" method"));
```
Comment (yegappan) [inline src/errors.h]: Can you change "use a object" to "use an object"?

File: src/testdir/test_vim9_class.vim [hunk 5]
```diff
@@ -955,6 +955,28 @@ def Test_class_new_with_object_member()
...
+  END
+  v9.CheckSourceFailure(lines, 'E1390:')
```
Comment (yegappan) [inline src/testdir/test_vim9_class.vim]: I started using complete error messages for verification as this makes the test easier to read and understand.  Can you include the complete the error message here?

**Review by yegappan — CHANGES_REQUESTED** [review]

**Comment by h-east** [comment]
@yegappan Thanks for the review!
I have accepted all suggestions and made the corrections.

**Review by yegappan — APPROVED** [review]

**Pull request closed without merging**